### PR TITLE
BLD: use tempita for hashtable

### LINKED ISSUE #13786: CLN: template hashtable.pyx
use Tempita support to provide templates for routines in ``hashtable.pyx``. Lots of duplicated code by dtype that can easily use fused types, so code generation is needed. Following along the lines of #13716 would be nice.

### PR DESCRIPTION
**Pull request by sinhrks**
- [x] closes #13786 
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`

### CODE REVIEW

File: pandas/src/hashtable_class_helper.pxi.in [hunk 1]
```diff
@@ -0,0 +1,642 @@
...
+
+        with nogil:
+            for i in range(n):
+                val = values[i]
+
+                {{if float_group}}
+                if val == val:
+                {{endif}}
+
+                {{tab}}k = kh_get_{{dtype}}(self.table, val)
```
Comment (shoyer) [inline pandas/src/hashtable_class_helper.pxi.in]: `{{tab}}` is pretty ugly! If tempita doesn't have a way to handle conditional indentation I would consider simply making separate blocks for each condition. That will be slightly more code but easier to understand.

**Comment by jreback** [comment]
be sure to add to the linter as well.

File: pandas/src/hashtable_class_helper.pxi.in (hunk 1 above)
Comment (sinhrks) [inline pandas/src/hashtable_class_helper.pxi.in]: Yes, using if-else is clearer in this case. fixed.

**Comment by sinhrks** [comment]
@jreback OK. linting is enabled for all ``src/*.pxi.in`` by default.

- https://github.com/pydata/pandas/blob/master/ci/lint.sh#L35

**Comment by jreback** [comment]
thanks @sinhrks I know you wanted to add other types (e.g. int32), so can make another issue for that, but this is great.

**Pull request merged**